### PR TITLE
Added new centrality metric

### PR DESCRIPTION
**Pull request by tibrewalpratik17**
Added Alpha Centrality Metric. The links which I referred to are:
* [Alpha Centrality](https://en.wikipedia.org/wiki/Alpha_centrality)
* [Katz Centrality](https://en.wikipedia.org/wiki/Katz_centrality)
* [Further reading on Katz Centrality](https://www.sci.unich.it/~francesc/teaching/network/katz.html)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
+/*
+ * (C) Copyright 2016-2018, by Dimitrios Michail and Contributors.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Copyright should be you as original author, with year 2018-2018 since the file is new.

**Comment by tibrewalpratik17** [comment]
Done @jsichi

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+ * iterations or if the AlphaCentrality values between two iterations do not change more than a predefined
+ * value.
+ * </p>
+ *
+ * <p>
+ * Each iteration of the algorithm runs in linear time O(n+m) when n is the number of nodes and m
+ * the number of edges of the graph. The maximum number of iterations can be adjusted by the caller.
+ * The default value is {@link AlphaCentrality#MAX_ITERATIONS_DEFAULT}.
+ * </p>
+ *
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: You might want to mention that the graph should not have negative edge weights.

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+import org.jgrapht.graph.*;
+
+import junit.framework.*;
+
+/**
+ * Unit tests for AlphaCentrality
+ * 
+ * @author Pratik Tibrewal
+ */
+public class AlphaCentralityTest
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: We recently got rid of all of the old JUnit3 test conventions and replaced them with JUnit4.  So for new code:

1) do use the `@Test` annotation

2) do NOT extend `TestCase`

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        }
+    }
+
+    public void testBadParameters()
+    {
+        DirectedPseudograph<String, DefaultEdge> g = new DirectedPseudograph<>(DefaultEdge.class);
+
+        try {
+            new AlphaCentrality<>(g, 1.1, 1.0, 100, 0.0001);
+            fail("No!");
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: Better would be to use the JUnit4 annotation `@Test(expected=IllegalArgumentException)`

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+ */
+package org.jgrapht.alg.scoring;
+
+import org.jgrapht.alg.interfaces.*;
+import org.jgrapht.graph.*;
+
+import junit.framework.*;
+
+/**
+ * Unit tests for AlphaCentrality
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: It looks like you took the example graphs verbatim from PageRankTest.  Shouldn't you be also adding some new graphs which demonstrate why one might want to use this measure?  Also, calling on the same input graph more than once with different parameters is a good way to demonstrate their effect.  (Unit tests serve an additional purpose of implicit documentation.)

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    {
+        this(g, dampingFactor, EXOGENOUS_FACTOR_DEFAULT, MAX_ITERATIONS_DEFAULT, TOLERANCE_DEFAULT);
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactor the exogenous factor
```
[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: So should I expect that user can enter any one of the above: like either a scalar or a map or I should go for only maps? Instead of vectors I was thinking of using maps as it will be faster. 
If we go for the implementation of both scalar/map, then I think it will be more flexible. If the user enters a scalar I can convert it into a map with all the values as that of the scalar or should I go for writing 2 run functions?

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java (hunk 5 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: I used the graph mentioned in this [page](https://www.sci.unich.it/~francesc/teaching/network/katz.html). It turns out that the implementation of the [PageRankTest graph](https://www.sci.unich.it/~francesc/teaching/network/pagerank) is same as that of I tested. I think the difference in the outputs of both the graphs demonstrates the difference in the implementations of both the graphs, isn't it?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Just converting the scalar to map would be fine.

[the remark this answers is not in the document]
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Done

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+     * @param maxIterations the maximum number of iterations to perform
+     * @param tolerance the calculation will stop if the difference of AlphaCentrality values between
+     *        iterations change less than this value
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, double exogenousFactor, int maxIterations, double tolerance)
+    {
+        this.g = g;
+        this.scores = new HashMap<>();
+
+        if (maxIterations <= 0) {
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Don't repeat the code for parameter validation.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+
+        if (dampingFactor < 0.0 || dampingFactor > 1.0) {
+            throw new IllegalArgumentException("Damping factor not valid");
+        }
+
+        if (tolerance <= 0.0) {
+            throw new IllegalArgumentException("Tolerance not valid, must be positive");
+        }
+
+        Map<V, Double> exofactor = new HashMap<>();
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Rather than actually building the map, you could change the run method to take a function as input.

Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Can you please explain me a bit more..? I didn't get you exactly..

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 7 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Can I do the parameter evaluation in the run method instead of the constructors...this way I will be evaluating parameters only once...

Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Better to share a validation method.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 8 above)
Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Run parameter type would be Function<Vertex, Double>.

For scalar, apply always returns the constant value.

For map, apply does the map lookup.

If this still isn't clear, you'll want to read up on Java 8 functions/lambda.

Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: What I think @jsichi means is you can convert exofactor from a `Map<V, Double>` to a `ToDoubleFunction<V>`.

So instead of:

```java
Map<V, Double> exofactor = ...
run(dampingFactor, exofactor, maxIterations, tolerance);

. . .

private void run(double dampingFactor, Map<V, Double> exofactormap, int maxIterations, double tolerance)
{
    ...
    double vNewValue = contribution + exofactormap.get(v);
    ...
}
```

perhaps something like:

```java
ToDoubleFunction<V> exofactorFunction = v -> exogenousFactor;
run(dampingFactor, exofactorFunction, maxIterations, tolerance);

. . .

private void run(double dampingFactor, ToDoubleFunction<V> exofactorFunction, int maxIterations, double tolerance)
{
    ...
    double vNewValue = contribution + exofactorFunction.applyAsDouble(v);
    ...
}
```

The advantage of this is it would allow the exofactors of each vertex to be computed only when they're really needed, which should save space and time compared to creating a map upfront.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactormap the exogenous factor map
+     * @param maxIterations the maximum number of iterations to perform
+     * @param tolerance the calculation will stop if the difference of AlphaCentrality values between
+     *        iterations change less than this value
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, Map<V, Double> exogenousFactormap, int maxIterations, double tolerance)
```
Comment (jbduncan) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Similarly to what I described over at https://github.com/jgrapht/jgrapht/pull/501/files#r172380977, I suspect it's desirable for `exogenousFactormap` to be a `ToDoubleFunction<V>`, but I'd wait for the JGraphT team to comment on it first before changing anything.

Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: "exogenousFactormap" -> "exogenousFactor"?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 8 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Well, I say "computing", but in this case the code is just retrieving an effectively constant exogenous factor, so no computation is involved per se, but it would still save space and time that it would spend otherwise creating a map. :)

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 7 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @jsichi Are you saying that @tibrewalpratik17 should extract the following code into a helper method, to prevent duplication across the two `AlphaCentrality` constructors?

```java
if (maxIterations <= 0) {
  throw new IllegalArgumentException("Maximum iterations must be positive");
}

if (dampingFactor < 0.0 || dampingFactor > 1.0) {
  throw new IllegalArgumentException("Damping factor not valid");
}

if (tolerance <= 0.0) {
  throw new IllegalArgumentException("Tolerance not valid, must be positive");
}
```

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 8 above)
Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Right, what he said.  :)

I spend 90% of my time in scala land these days, so when I mention anything from Java 8 I am waving my hands.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: What I have done is I have used a function to look up to the map for fetching a value particularly related to a vertex. 
`ToDoubleFunction<Integer> i  = (x)-> f(x);` here your `f(x)` is the task you want to do when you receive x. So in our case, we have `V` instead of `Integer` as we are taking a vertex as input and so in case of scalar my `f(v)` is just the `exogenousfactor` and in case of map `f(v)` is `exogenousFactormap.get(v)` which corresponds to the exogenous factor value of that particular vertex.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 8 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Done @jsichi. Moreover should I squash these 2 new commits or it is fine?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Ah, okay, I think I see what you've done.

My knowledge of alpha centrality is virtually nonexistent, so I don't have the full context here, but I still wonder though if it would be worth converting the `Map<V, Double>` parameter in this constructor to a `ToDoubleFunction<V>`.

The way I see it, it would give a user a lot more flexibility to pass in any way of converting vertices to exogenous factors that they can think of (so they aren't limited to just maps), and AFAICT it's much easier to convert a map to a function than the other way round.

(Compare:

```java
ToDoubleFunction<V> f = v -> ourExogenousFactorMap.get(v);
new AlphaCentrality(..., f, ...);
```

with

```java
Map<V, Double> m = ourVertices.stream().collect(toMap(v -> v, ourExogenousFactorFunction));
new AlphaCentrality(..., m, ...);
// or
Map<V, Double> m = ourVertices.stream().collect(toMap(v -> v, ourExogenousFactorFunction, (a, b) -> b, HashMap::new));
new AlphaCentrality(..., m, ...);
```

I'd argue the first option is more readable).

Not to mention that a `Map<V, Double>` would necessarily take more space and one is forced to box primitive `double`s to `Double` objects. :)

Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @jbduncan The user does not want to convert vertices to exogenous factors but they want to assign each vertex a separate factor or they assign all the vertices the same factor by passing a scalar. Basically we are already giving the user a lot of flexibility by allowing them to pass a scalar or a map instead of converting the scalar to a map himself/herself.

Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Okay. I admit that I'm not convinced by this argument, so I think it's best if we let the JGraphT team decide which route we should go. :)

Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @jbduncan I think there is always a trade-off between certain things...in our case be it space for creating maps or allowing user to enter parameters other than maps. But I think if we refer to other standard libraries then they have implemented this particular algorithm for maps and scalars only. Moreover I think maps being fast should be preferred from user's point of view too.
I am really sorry if I am not able to convince you till now but would love to hear your points too.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+     * @param maxIterations the maximum number of iterations to perform
+     * @param tolerance the calculation will stop if the difference of AlphaCentrality values between
+     *        iterations change less than this value
+     */
+    public AlphaCentrality(Graph<V, E> g, double dampingFactor, double exogenousFactor, int maxIterations, double tolerance)
+    {
+        this.g = g;
+        this.scores = new HashMap<>();
+
+        Map<V, Double> exofactor = new HashMap<>();
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: I think the garbage man forgot to empty the trash bins today?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: I tend to agree with @jbduncan.  Since JGraphT has been around since long before Java 8, a lot of our code does not take advantage of functional programming capabilities, but for new code, it's worth fresh thinking.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+    private void run(double dampingFactor, ToDoubleFunction<V> exofactorFunction, int maxIterations, double tolerance)
+    {
+        // initialization
+        int totalVertices = g.vertexSet().size();
+
+        double initScore = 1.0d / totalVertices;
+        for (V v : g.vertexSet()) {
+            scores.put(v, initScore);
+            }
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: weird indentation

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        if (dampingFactor < 0.0 || dampingFactor > 1.0) {
+            throw new IllegalArgumentException("Damping factor not valid");
+        }
+
+        if (tolerance <= 0.0) {
+            throw new IllegalArgumentException("Tolerance not valid, must be positive");
+        }
+        return true;
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: What is the point of this return value?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @tibrewalpratik17 No worries!

Warning, and apologies in advance: wall of text incoming!

The only other argument I can think of in favour of `ToDoubleFunction<V>` over `Map<V, Double>` is that is better captures the semantics of what we're trying to do with this map.

To explain what I mean, in the `run` method of `AlphaCentrality`, the most that we're doing with the `ToDoubleFunction<V>` param there is we're querying it for an exogenous factor for a given vertex. In that sense, we're only really using the `Map#get` method of the passed-in map, rather than say `Map#put` as well. I don't imagine after this has all been reviewed that there'll be a need to depend on other methods of `Map` for what is effectively a function that accepts vertices and returns exogenous factors. Thus, the way I see it, if we tell the user through the API that they need to pass in a `java.util.Map`, then we're potentially saying that we need not just `Map#get` but other methods on `Map` as well, which AFAICT we don't.

So if a user has some other similar but sufficiently different data structure e.g. a [Vavr `Map`](https://www.javadoc.io/doc/io.vavr/vavr/0.9.2), then they've no choice but to convert their existing data structure to a `java.util.Map`, which may be costly in terms of either runtime performance or the mental effort needed to implement the code to do the conversion. Making the `Map` parameter a `ToDoubleFunction` would fix this problem.

Guava's graphs package approaches things similarly where for algorithms that only need to use one method of an input graph to do its job, e.g. `Graph#successors` for breadth first search, then the algorithm accepts e.g. a `SuccessorFunction` instead of a full-blown `Graph`, which allows e.g. users of JGraphT graphs to use Guava's graph algorithms. (I think this link to the [Guava wiki](https://github.com/google/guava/wiki/GraphsExplained#graph-types) explains things a bit better than I've done here.)

But having said all this, going back to the Vavr example, Vavr's `Map` does have a `toJavaMap` method which I believe is relatively efficient, so that particular example is a bit contrived.

**tl;dr:** `ToDoubleFunction` is more focused API-wise than `Map` since it has only one method, rather than N methods some of which could be used to mutate the passed-in `Map`, which we don't want to do.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 10 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Hehe.. Sorry!!

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 11 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: rectifying...

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 12 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Yup can surely get rid of it :)

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @jbduncan I went through the docs of [Successor Functions](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/graph/SuccessorsFunction.html) and [Predecessor Functions](http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/graph/PredecessorsFunction.html) and really understood your point.

Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @jsichi What will you suggest regarding it's implementation in this class?

Comment (jsichi) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: What @jbduncan suggested:  change the current constructor which takes a Map into one which takes a ToDoubleFunction.

Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Done @jsichi.
Thanks @jbduncan. After implementing what you said I really got what exactly you were trying to point out. Basically converting a map or other data structures to a `ToDoubleFunction` is rather simple but vice-versa is not. Thus the user can simply convert his/her data structure into a `ToDoubleFunction` and pass as a parameter rather than to look for converting it into `Maps`. Sorry for all the trouble and I hope I got what you were saying this time correctly. Thanks a lot!!

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+        ToDoubleFunction<V> exofactorFunction = (v) -> exogenousFactor;
+        run(dampingFactor, exofactorFunction, maxIterations, tolerance);
+    }
+
+    /**
+     * Create and execute an instance of AlphaCentrality.
+     * 
+     * @param g the input graph
+     * @param dampingFactor the damping factor
+     * @param exogenousFactorFunction ToDoubleFunction which can contain function related to maps
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: you could just say "a provider of exogenous factors per vertex"

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+        DirectedPseudograph<String, DefaultEdge> g = new DirectedPseudograph<>(DefaultEdge.class);
+
+        g.addVertex("1");
+        g.addVertex("2");
+        g.addEdge("1", "2");
+        g.addEdge("2", "1");
+
+        VertexScoringAlgorithm<String, Double> pr = new AlphaCentrality<>(g);
+
+        // assertEquals(pr.getVertexScore("1"), pr.getVertexScore("2"), 0.0001);
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: Why is this commented out?

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 13 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: Yup this sounds better!

File: jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java (hunk 14 above)
Comment (tibrewalpratik17) [reply jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: Fixed! Moreover @jsichi do you recommend squashing these commits or is it fine? I think squashing will be good right?

Comment (jsichi) [reply jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: Yes, squash please.

Comment (tibrewalpratik17) [reply jgrapht-core/src/test/java/org/jgrapht/alg/scoring/AlphaCentralityTest.java]: Done.

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 13 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: 👍

File: jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java (hunk 9 above)
Comment (jbduncan) [reply jgrapht-core/src/main/java/org/jgrapht/alg/scoring/AlphaCentrality.java]: @tibrewalpratik16 You're welcome! 😀

**Comment by tibrewalpratik17** [comment]
@jsichi I wanted to ask how good it will be to add a class for Eigenvector Centrality considering the fact that Alpha Centrality with exogenous factor as zero is the same as Eigenvector Centrality.. So as a package we should focus on providing all the algorithms or we should also depend on the user's knowledge that he knows to implement Eigenvector Centrality he may use Alpha Centrality too..?

**Comment by tibrewalpratik17** [comment]
Done. Yes but at first I am thinking of implementing some other centrality metrics then we can look into linear algebra too :)

**Comment by jsichi** [comment]
Thanks Pratik, I'm merging this one now.
